### PR TITLE
[WIP] Lexers improvement

### PR DESCRIPTION
**Pull request by LoicGrobol**
The original intent behind this PR was to make the BERT lexer give its BERT layer inputs that are closer to what it has been trained with and get as much information as possible from its output. Along the way, many other changes have been done where it seemed to make sense or to improve the quality of the word representations.

The changes tend to have a clear positive impact on the parser accuracy. However,  break backward compatibility with the old models in ways that are not easily circumvented, so they will all need retraining (which will, however give the opportunity to have more accurate models) but might imply the need for archival of the code at its present state, e.g. to keep the original FlauBERT paper reproducible.

Notes:

- The FastText embeddings are now trainable by default, which seems to improve the performance slightly when also using a BERT model, but degrades it without. More investigation needed for this and probably hyperparameters tuning too.
- The embeddings of the root in the FastText and BERT embeddings are set to have distributions resembling those of the embedding of non-root words. Since the root token is the most frequent token in the corpus, it could be (and in practice often is)  harmful to e.g. set it to zero. However, the theoretical motivations of the choices made here are flimsy and more investigation in that direction could be beneficial.

## Changed

- The BERT embeddings have undergone drastic changes:
  - BERT is now passed on the full sequence of word pieces, and not the sequence of the first wordpiece of each word
  - The special BOS and EOS tokens are now added, making the input closer to what BERT was trained with
  - The BERT embedding of a word is now either the embedding of its first wordpiece (with the config option `bert_subwords_reduction: "first`, which is also the default if the option is absent) or the mean of the embeddings of its wordpiece (with the config option `bert_subwords_reduction: "mean`).
- Embeddings for the root token have changed in several lexers
  - In the char and FastText lexers, the root token is now considered as a single-character “special” token with a dedicated embedding, in both case trainable. In both lexers, it was previously decomposed in individual characters `"<"`, `"r"`, `"o"`…
  - The BERT part of `BertBaseLexer` now use the unweighted average of the embeddings of all subwords  for the root token.
- FastText embeddings are now *trainable* by default. The old behaviour is available with a config option (see below).
- The FastText embedding of a word is now the mean of the embedding of its subwords instead of the sum, which was making the magnitude of the embedding dependent on the length of the word. Additionally we checked that the padding does not actually enter this mean (since the padding embeddings are kept at zero).
- `BertBaseLexer` and `DefaultLexer` are now responsible for batching and padding their inputs via a `pad_batch` method that returns something acceptable as an input to their `forward` method, making it all transparent from the `DependencyDataset` point of view and allowing a drastic simplification of `DependencyDataset.pad`.
- For all lexers that use it, word dropout is now done in `forward` rather than in tokenization and is now done directly on the tensorized inputs, which is faster, although it has the minor side effect of turning some of the padding into `<unk>`s.
- `DependencyDataset` shuffling and ordering is now done exclusively when building batches and does not affect the internal order, which removes the need to re-encode the trees at every epoch.
- Lexers that use specific unk, pad or root indices or tokens now require them in their constructor instead of statically importing them from `DepTree` or `DependencyDataset`, which helps decouple `deptree` and `lexers` and did not seem to much of a hassle in practice.
- Some functions that used to return bare tuples now return typed namedtuples, which has no other side effect, but improves the typing support and could in the future be used to simplify other parts of the code.
- The biaffine layer is now an actual biaffine layer (with the linear and bias parts) as in the original Dozat and Manning version instead of simply bilinear (the old behaviour is now behind an hyperparameter, see below). It has also been rewritten to be faster and more memory efficient, which somewhat counterbalance the loss of efficiency in the BERT lexer.

## Removed

- The `cased` hyperparameter for BERT configs is now ignored: uncased BERT tokenizers now lowercase their inputs anyway so we have no need to do it for them and the corner case of using a cased BERT tokenizer while passing it lowercased inputs seems to marginal to keep that code path.
- `FastTextTorch.train_model_from_trees` has been removed, replaced by `FastTextTorch.train_model_from_sents` which takes a list of lists of words instead of a list of trees. The old behaviour can be achieved with `FastTextTorch.train_model_from_sents([tree.words[1:] for tree in traintrees], fasttext_model_path)`.

## Added

- A `freeze_fasttext` boolean hyperparameter that mirrors the `freeze_bert` and can be set to true to disable the fine-tuning of FastText embeddings.
- A `biased_biaffine` boolean hyperparameter that can be set to false to revert to the old behaviour of making the biaffine layer actually simply bilinear.

### CODE REVIEW

**Comment by LoicGrobol** [comment]
Closing to go around a github bug

**Comment by bencrabbe** [comment]
I think you can merge.

**Comment by LoicGrobol** [comment]
I couldn't get it to find the last commit so I had to close and open #16 >.<

**Comment by LoicGrobol** [comment]
Which I think I'll merge today with the GSD models and the FTB models (I have the spmrl version but I wanted to try on the UD version too and I don't have it yet) will wait.

**Comment by LoicGrobol** [comment]
Also I meat to write model cards for these but still haven't got around that… eh, later
